source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,875 @@
+to the application. Drivers MUST document that users can either call ``next``
+again or close the existing change stream and create a new one if a previous
+``next`` call times out. The documentation MUST suggest closing and
+re-creating the stream with a higher timeout if the timeout occurs before any
+events have been received because this is a signal that the server is timing
+out before it can finish processing the existing oplog.
+
+See `Change stream behavior`_.
+
+Sessions
+--------
+
+The `SessionOptions <../sessions/driver-sessions.rst#mongoclient-changes>`_
+used to construct explicit `ClientSession
+<../sessions/driver-sessions.rst#clientsession>`_ instances MUST accept a new
+``defaultTimeoutMS`` option, which specifies the ``timeoutMS`` value for the
+following operations executed on the session:
+
+#. commitTransaction
+#. abortTransaction
+#. withTransaction
+#. endSession
+
+If this option is not specified for a ``ClientSession``, it MUST inherit the
+``timeoutMS`` of its parent MongoClient.
+
+Convenient Transactions API
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If ``timeoutMS`` is set, drivers MUST apply it to the entire
+``withTransaction`` call. To propagate the timeout to the user-supplied
+callback, drivers MUST store the timeout as a field on the ClientSession
+object. This field SHOULD be private to ensure that a user can not modify it
+while a ``withTransaction`` call is in progress. Drivers that cannot make
+this field private MUST signal that the field should not be accessed or
+modified by users if there is an idiomatic way to do so in the language (e.g.
+underscore-prefixed variable names in Python) and MUST document that
+modification of the field can cause unintended correctness issues for
+applications. Drivers MUST document that the remaining timeout will not be
+applied to callback operations that do not use the ClientSession and that
+overridding ``timeoutMS`` for operations inside the provided callback will
+result in a client-side error. If the callback returns an error and the
+transaction must be aborted, drivers MUST refresh the ``timeoutMS`` value for
+the ``abortTransaction`` operation.
+
+If ``timeoutMS`` is not set, drivers MUST continue to exhibit the existing
+120 second timeout behavior. Drivers MUST NOT change existing implementations
+to use ``timeoutMS=120000`` for this case.
+
+See `withTransaction communicates timeoutMS via ClientSession`_ and
+`withTransaction refreshes the timeout for abortTransaction`_.
+
+GridFS API
+----------
+
+GridFS buckets MUST inherit ``timeoutMS`` from their parent MongoDatabase
+instance and all methods in the GridFS Bucket API MUST support the
+``timeoutMS`` option. For methods that create streams (e.g.
+``open_upload_stream``), the option MUST cap the lifetime of the entire
+stream. This MUST include the time taken by any operations executed during
+stream construction, reads/writes, and close/abort calls. For example, if a
```

Comment:
Coming back to this after reading "GridFS streams behavior". Is this actually talking about tracking elapsed wall time (even outside of GridFS methods)? If so, does that mean that a user could hit a timeout error in the following scenario:

* User opens a GridFS stream with `timeoutMS=1000`
* User spends 5 seconds doing tasks unrelated to the driver
* User attempts to write data to GridFS and encounters a timeout error because 5+ seconds have already elapsed

Reply:
The scenario you described would cause a user to hit the timeout. We struggled to define the GridFS semantics for this project, but eventually decided that users are _likely_ not opening streams and leaving them around while they do other work. Obviously, it's possible that this assumption is wrong, but AFAIK, GridFS is not widely used, so I think we can come back to this decision if we see users running into problems with it.

Reply:
If we decide the current behavior (`timeoutMS` counts the in-memory lifetime of the stream), would changing this behavior be considered a BC break? If so, I'm not optimistic many drivers would get around to changing it. Or is there a possibility we'd collect early user feedback on this from beta testers?

Since we can go either way on this, I guess it comes down to asking which implementation leaves users worse off if we get it wrong and can't easily change behavior? Would users be surprised to learn that `timeoutMS` applies to general clock time that the stream stays open in memory? And even if we changed that behavior and just limited `timeoutMS` to tracking time spent within GridFS methods, would they be surprised to learn it applied to time spent read/writing the non-GridFS stream?

Side note: if any user happens to use `timeoutMS` at a higher application level (to generally apply to client/database ops) and also uses GridFS, I expect they'll end up needing to override `timeoutMS` for GridFS by either disabling it or specifying a much greater timeout. Perhaps we might expect the same for folks using `bulkWrite` and `withTransaction`.

Reply:
Given the size of this PR, I'll move this discussion offline into Slack so others can weigh in as well.

Reply:
I think this can be resolved. I see you added "Modify GridFS streams behavior via new options" under "Future Work".